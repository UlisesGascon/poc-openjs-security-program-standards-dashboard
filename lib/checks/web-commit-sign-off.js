export function checkWebCommitSignOff (projects, checks) {
  checks.webCommitSignOff = {
    name: 'Web Commit Sign-off',
    description: 'Github Org Requires Commit Sign-off for Web-Based Commits',
    level: {
      incubating: 'recommended',
      graduated: 'recommended',
      retiring: 'recommended'
    },
    url: 'https://openjs-security-program-standards.netlify.app/details/item-62',
    projects: {}
  }

  for (const project of projects) {
    checks.webCommitSignOff.projects[project.name] = {}

    let totalRepos = 0
    let nonCompliantRepos = 0
    for (const gitHubOrg of project.githubOrgs) {
      totalRepos += gitHubOrg.repositories.length
      nonCompliantRepos = gitHubOrg.repositories.filter(repo => !repo.web_commit_signoff_required).length

      if (nonCompliantRepos > 0) {
        const percentage = (nonCompliantRepos / totalRepos) * 100
        checks.webCommitSignOff.projects[project.name].globalAchievement = false
        checks.webCommitSignOff.projects[project.name].rationale = `There are ${nonCompliantRepos} (${percentage.toFixed(2)}%) repositories that do not require commit sign-off for web-based commits.`
      } else {
        checks.webCommitSignOff.projects[project.name].globalAchievement = true
        checks.webCommitSignOff.projects[project.name].rationale = 'All repositories require commit sign-off for web-based commits.'
      }
    }
  }
}
