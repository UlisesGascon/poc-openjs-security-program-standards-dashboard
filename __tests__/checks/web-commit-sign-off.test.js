import { checkWebCommitSignOff } from '../../lib/checks/web-commit-sign-off.js'

describe('checkWebCommitSignOff', () => {
  let projects
  let checks
  let tasks
  let alerts

  beforeEach(() => {
    projects = [
      {
        name: 'Project1',
        githubOrgs: [
          {
            repositories: [
              { 
                web_commit_signoff_required: true,
                full_name: 'org1/repo',
                html_url: 'https://github.com/org1/repo'
              },
              { 
                web_commit_signoff_required: false,
                full_name: 'org1/repo2',
                html_url: 'https://github.com/org1/repo2'
              }
            ]
          }
        ]
      },
      {
        name: 'Project2',
        githubOrgs: [
          {
            repositories: [
              { 
                web_commit_signoff_required: true,
                full_name: 'org2/repo',
                html_url: 'https://github.com/org1/repo'
              },
              { 
                web_commit_signoff_required: true,
                full_name: 'org2/repo2',
                html_url: 'https://github.com/org1/repo2'
              }
            ]
          }
        ]
      }
    ]
    checks = {}
    tasks = {}
    alerts = {}
  })

  it('should update checks with project compliance data', () => {
    checkWebCommitSignOff({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with no repositories', () => {
    projects = [
      {
        name: 'Project3',
        githubOrgs: [
          {
            repositories: []
          }
        ]
      }
    ]
    checkWebCommitSignOff({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })
})