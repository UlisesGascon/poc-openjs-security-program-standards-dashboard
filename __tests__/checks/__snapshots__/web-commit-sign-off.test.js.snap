// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkWebCommitSignOff should handle projects with no repositories 1`] = `
{
  "webCommitSignOff": {
    "category": "source control",
    "description": "Github Org Requires Commit Sign-off for Web-Based Commits",
    "level": {
      "graduated": "recommended",
      "incubating": "recommended",
      "retiring": "recommended",
    },
    "name": "Web Commit Sign-off",
    "projects": {
      "Project3": {
        "globalAchievement": true,
        "rationale": "All repositories require commit sign-off for web-based commits.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-62",
  },
}
`;

exports[`checkWebCommitSignOff should handle projects with no repositories 2`] = `{}`;

exports[`checkWebCommitSignOff should handle projects with no repositories 3`] = `{}`;

exports[`checkWebCommitSignOff should update checks with project compliance data 1`] = `
{
  "webCommitSignOff": {
    "category": "source control",
    "description": "Github Org Requires Commit Sign-off for Web-Based Commits",
    "level": {
      "graduated": "recommended",
      "incubating": "recommended",
      "retiring": "recommended",
    },
    "name": "Web Commit Sign-off",
    "projects": {
      "Project1": {
        "globalAchievement": false,
        "rationale": "There are 1 (50.00%) repositories that do not require commit sign-off for web-based commits.",
      },
      "Project2": {
        "globalAchievement": true,
        "rationale": "All repositories require commit sign-off for web-based commits.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-62",
  },
}
`;

exports[`checkWebCommitSignOff should update checks with project compliance data 2`] = `
{
  "Project1": {
    "webCommitSignOff": {
      "category": "source control",
      "description": "Enable commit sign-off for web-based commits in all repositories",
      "name": "Enable Commit Sign-off for Web-Based Commits",
      "priority": "low",
      "repositories": [
        {
          "full_name": "org1/repo2",
          "html_url": "https://github.com/org1/repo2",
        },
      ],
      "url": "https://openjs-security-program-standards.netlify.app/details/item-62",
    },
  },
}
`;

exports[`checkWebCommitSignOff should update checks with project compliance data 3`] = `{}`;
